### PR TITLE
SQL Server Driver :yum:

### PR DESCRIPTION
**Pull request by camsaul**
Implements #1298

### CODE REVIEW

**Comment by camsaul** [comment]
Tests are passing, Christmas is coming, the goose is getting fat. The SQL Server driver is officially ready to go (IMO) :yum: 

Quite a lot of code improvement and refactoring happened along the way. There was a pretty big refactor of the code for creating and destroying temporary test databases (mostly confined to the SQL drivers). The the code for loading temporary SQL databases for individual DB engines is substantially simplified. `metabase.test.data.postgres.clj`, for example, is now about 50 lines compared to over 100. And tweaking any individual part to do something needed specifically by a single DB engine is pretty easy now too. :scream_cat: 

On that note, I think I finally figured out a way to have our cake and eat it too with regards to Clojure protocols:

1.  Define absolutely every function and property related to the task at hand in as part of one mega-interface. Break apart large functions as much as possible. `IGenericSQLDatasetLoader` now defines around 16 methods. :heart_eyes_cat: 
2.  Create default implementations for as many of these functions as humanly possible and *package them in a mixin*. Use `defn` to declare the functions in case they need to be called outside of the mixin. `metabase.test.data.generic-sql/DefaultsMixin` provides default implementations for 14 of the 16 methods in `IGenericSQLDatasetLoader`. :sob:
3.  Types that want to implement a protocol should use `extend` and `merge` in the mixin, implementing any methods that don't have defaults and overriding any ones in cases where they need to do something wacky. Since they have access to the default implementations, they can still call those if needed (analogous to calling a method on a superclass) :smiling_imp: 
4.  DOPE AS :moneybag: Now implementing things is easy. Hairy special case code is easy to inject in the exact spots where you need it and is confined to a specific DB engine rather than being baked into some core logic. Adding new *optional* methods to a protocol is easy, since everyone implementing the protocol is "inheriting" your defaults mixin :100: 

`</enlightenment>`

I see MB drivers working like this in the near future. One *GIANT* `IDriver` protocol with 80+ methods, 70 of which would probably have default implementations (70 of which would be overridable to do special-case things specific to the DB engine from hell of the week *cough SQL Server*). We're already doing something pretty similar with my last refactor, just without the formality of a protocol.

TL;DR

SQL Server driver is done :+1: :heart_eyes_cat: :yum:

File: test/metabase/driver/query_processor_test.clj [hunk 1]
```diff
@@ -570,7 +572,7 @@
 
 
 ;;; ### order_by aggregate ["sum" field-id]
-(qp-expect-with-datasets #{:h2 :postgres :mysql}
+(qp-expect-with-datasets #{:h2 :postgres :mysql :sqlserver}
```
Comment (agilliland) [inline test/metabase/driver/query_processor_test.clj]: should we define this set once and reference it throughout this file?

Comment (camsaul) [inline test/metabase/driver/query_processor_test.clj]: Good idea. I just pushed that change

File: src/metabase/driver/generic_sql.clj [hunk 2]
```diff
@@ -156,10 +157,18 @@
...
       A map of native DB column types (as keywords) to the `Field` `base-types` they map to.
 
-   *  `sql-string-length-fn`
+   *  `string-length-fn`
 
       Keyword name of the SQL function that should be used to get the length of a string, e.g. `:LENGTH`.
 
+   *  `stddev-fn` *(OPTIONAL)*
+
+      Keyword name of the SQL function that should be used to get the length of a string. Defaults to `:STDDEV`.
```
Comment (agilliland) [inline src/metabase/driver/generic_sql.clj]: this looks like copy pasta :spaghetti:

File: src/metabase/driver/sqlserver.clj [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+   :xml              :UnknownField
+   (keyword "int identity") :IntegerField}) ; auto-incrementing integer (ie pk) field
+
+(defn- connection-details->spec [details]
+  (-> (kdb/mssql details)
+      ;; swap out Microsoft Driver details for jTDS ones
+      (assoc :classname   "net.sourceforge.jtds.jdbc.Driver"
+             :subprotocol "jtds:sqlserver")
+      ;; adjust the connection URL to match up with the jTDS format (see http://jtds.sourceforge.net/faq.html#urlFormat)
+      (update :subname #(s/replace % #";database=" "/"))))
```
Comment (agilliland) [inline src/metabase/driver/sqlserver.clj]: this seems like it could be dangerous because we are assuming that `;database=` is the first argument on the connection string.  is that something that's guaranteed?

Comment (camsaul) [inline src/metabase/driver/sqlserver.clj]: Yup we pass in the details map and that's the string korma makes out of it. So unless a future update to korma breaks things we're ok. In that case our tests would fail when we tried to bump dependencies

File: src/metabase/driver/generic_sql.clj (hunk 2 above)
Comment (camsaul) [inline src/metabase/driver/generic_sql.clj]: There probably is some way to get the length of a string with STDDEV :disappointed:
